lib/TransactionRunner.js

```diff
@@ -42,19 +42,43 @@ class TransactionRunner {
   }
 
   run(transactions, callback) {
-    logger.debug('Sorting HTTP transactions');
-    transactions = this.configuration.options.sorted ? sortTransactions(transactions) : transactions;
+    logger.debug('Starting reporters and waiting until all of them are ready');
+    this.emitStart((emitStartErr) => {
+      if (emitStartErr) { return callback(emitStartErr); }
 
-    logger.debug('Configuring HTTP transactions');
-    transactions = transactions.map(this.configureTransaction.bind(this));
+      logger.debug('Sorting HTTP transactions');
+      transactions = this.configuration.options.sorted ? sortTransactions(transactions) : transactions;
 
-    // Remainings of functional approach, probs to be eradicated
-    logger.debug('Reading hook files and registering hooks');
-    addHooks(this, transactions, (addHooksError) => {
-      if (addHooksError) { return callback(addHooksError); }
+      logger.debug('Configuring HTTP transactions');
+      transactions = transactions.map(this.configureTransaction.bind(this));
 
-      logger.debug('Executing HTTP transactions');
-      this.executeAllTransactions(transactions, this.hooks, callback);
+      logger.debug('Reading hook files and registering hooks');
+      addHooks(this, transactions, (addHooksError) => {
+        if (addHooksError) { return callback(addHooksError); }
+
+        logger.debug('Executing HTTP transactions');
+        this.executeAllTransactions(transactions, this.hooks, (execAllTransErr) => {
+          if (execAllTransErr) { return callback(execAllTransErr); }
+
+          logger.debug('Wrapping up testing and waiting until all reporters are done');
+          this.emitEnd(callback);
```

Comment:
The `emitStart` and `emitEnd` methods were originally part of Dredd.js, but I moved them to the transaction runner. All other events are emitted in the transaction runner, so I thought it's a good idea to streamline the concerns.

Reply:
The code here is original, only start/end is added and the code is indented as a result of that.